src/main/downloadsManager.ts

```diff
@@ -0,0 +1,463 @@
+                item.cancel();
+                return;
+            }
+        } else {
+            const filename = this.createFilename(item);
+            const savePath = this.getSavePath(`${Config.downloadLocation}`, filename);
+            item.setSavePath(savePath);
+        }
+        this.upsertFileToDownloads(item, 'progressing');
+        this.progressingItems.set(this.getFileId(item), item);
+        this.handleDownloadItemEvents(item, webContents);
+        this.openDownloadsDropdown();
+        this.toggleAppMenuDownloadsEnabled(true);
+    };
+
+    /**
+     * This function monitors webRequests and retrieves the total file size (of files being downloaded)
+     * from the custom HTTP header "x-uncompressed-content-length".
+     */
+    webRequestOnHeadersReceivedHandler = (details: Electron.OnHeadersReceivedListenerDetails, cb: (headersReceivedResponse: Electron.HeadersReceivedResponse) => void) => {
+        const headers = details.responseHeaders ?? {};
+
+        if (headers?.['content-encoding']?.includes('gzip') && headers?.['x-uncompressed-content-length'] && headers?.['content-disposition'].join(';')?.includes('filename=')) {
+            const filename = readFilenameFromContentDispositionHeader(headers['content-disposition']);
+            const fileSize = headers['x-uncompressed-content-length']?.[0] || '0';
+            if (filename && (!this.fileSizes.has(filename) || this.fileSizes.get(filename)?.toString() !== fileSize)) {
+                this.fileSizes.set(filename, fileSize);
+            }
+        }
+
+        // With no arguments it uses the same headers
+        cb({});
+    };
+
+    checkForDeletedFiles = () => {
+        log.debug('DownloadsManager.checkForDeletedFiles');
+
+        const downloads = this.downloads;
+        let modified = false;
+
+        for (const fileId in downloads) {
+            if (Object.prototype.hasOwnProperty.call(downloads, fileId)) {
+                const file = downloads[fileId];
+                if ((file.state === 'completed')) {
+                    if (!file.location || !fs.existsSync(file.location)) {
+                        downloads[fileId].state = 'deleted';
+                        modified = true;
+                    }
+                } else if (file.state === 'progressing') {
+                    downloads[fileId].state = 'interrupted';
+                    modified = true;
+                }
+            }
+        }
+
+        if (modified) {
+            this.saveAll(downloads);
+        }
+    }
+
+    clearDownloadsDropDown = () => {
+        log.debug('DownloadsManager.clearDownloadsDropDown');
+
+        this.saveAll({});
+        this.fileSizes = new Map();
+        this.closeDownloadsDropdown();
+        this.toggleAppMenuDownloadsEnabled(false);
+    }
+
+    showFileInFolder = (item?: DownloadedItem) => {
+        log.debug('DownloadsDropdownView.showFileInFolder', {item});
+
+        if (item && fs.existsSync(item.location)) {
```

Comment:
Do we have a mechanism to mark the file deleted if it no longer exists? Might be something we want to do.

Reply:
What about after the app has started?

Reply:
If there's an update when it is clicked so that the user is shown that immediately, then we're okay. If the dropdown is opened, it's feasible that the user could open it, then go delete a file and then try and click on it afterwards.